hazelcast/src/main/java/com/hazelcast/core/FlakeIdGenerator.java

```diff
@@ -0,0 +1,81 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.core;
+
+import com.hazelcast.concurrent.flakeidgen.FlakeIdNodeIdOutOfRangeException;
+import com.hazelcast.internal.cluster.ClusterService;
+
+/**
+ * Creates cluster-wide unique ID generator. Generated IDs are {@code long} primitive values
+ * and are k-ordered (roughly ordered). IDs are in the range from {@code Long.MIN_VALUE} to
+ * {@code Long.MAX_VALUE}.
+ * <p>
+ * The IDs contain timestamp component and a node ID, which is assigned when the member
+ * joins the cluster. This allows the IDs to be ordered and unique without any coordination between
+ * members, which makes the generator safe even in split-brain scenario (for caveats,
+ * {@link ClusterService#getMemberListJoinVersion() see here}).
+ * <p>
+ * Timestamp component is in milliseconds since 1.1.2017, 0:00 GMT, and has 42 bits. This caps
+ * the useful lifespan of the generator to little less than 140 years. The sequence component is 6 bits.
+ * If more than 64 IDs are requested in single millisecond, IDs will gracefully overflow to next
+ * millisecond and uniqueness is guaranteed in this case.
+ *
+ * <h4>Performance</h4>
+ * Operation on member is typically local. On client, the {@link #newId()} method goes to a random
+ * member and gets a batch of IDs, which will then be returned locally for limited time. The pre-fetch
+ * size and the validity time can be configured for each client and member, see {@link
+ * com.hazelcast.config.Config#addFlakeIdGeneratorConfig(com.hazelcast.config.FlakeIdGeneratorConfig)
+ * here} for member config and {@code ClientConfig.addFlakeIdGeneratorConfig()} for client config.
+ *
+ * <h4>Node ID overflow</h4>
+ * Node ID component of the ID has 16 bits. Members with member list join version higher than
+ * 2^16 won't be able to generate IDs, but functionality will be preserved by forwarding to another
+ * member. It is possible to generate IDs on any member or client as long as there is at least one
+ * member with join version smaller than 2^16 in the cluster. The remedy is to restart the cluster:
+ * nodeId will be assigned from zero again. Uniqueness after the restart will be preserved thanks to
+ * the timestamp component.
+ */
+public interface FlakeIdGenerator extends DistributedObject {
+
+    /**
+     * Generates and returns a batch of cluster-wide unique IDs.
+     *
+     * @param batchSize Requested number of IDs.
+     * @return Iterable batch of unique IDs
+     *
+     * @throws FlakeIdNodeIdOutOfRangeException
+     *   if node ID for all members in the cluster is out of valid range.
+     * See "Node ID overflow" in {@link FlakeIdGenerator class documentation} for more details.
+     */
+    IdBatch newIdBatch(int batchSize);
```

Comment:
why do we have such a method in the interface? I think this should stay as internal logic and the only method should be `newId()`, as in the current `IdGenerator` interface.

Reply:
I asked the same question - https://github.com/hazelcast/hazelcast/pull/11864#discussion_r153730191

Reply:
I think it adds some useful flexibility. Removing it will not simplify the code much. If we do it for the sake of simplicity of the API, then I'm fine with it.

Reply:
I understand your point, but I think `newId()` is already enough from user's perspective. lets keep it private until it is explicitly requested by users.